spec/lib/metasploit/framework/credential_collection_spec.rb

```diff
@@ -323,6 +339,240 @@
+                Metasploit::Framework::Credential.new(public: "test_public2", private: ""),
+                Metasploit::Framework::Credential.new(public: "test_public1", private: nil),
+                Metasploit::Framework::Credential.new(public: "test_public2", private: nil),
+              )
+            end
+          end
+        end
+
+        context "when given a user file" do
+          let(:user_file) do
+            filename = "user_file"
+            stub_file = StringIO.new("asdf\njkl\n")
+            allow(File).to receive(:open).with(filename, /^r/).and_return stub_file
+
+            filename
+          end
+
+          # REMOVE BEFORE COMMIT: this also yields the usernames as passwords for the additional_public
+          context "when given a password" do
+            let(:password) { "password" }
+
+            specify  do
+              expect { |b| collection.each(&b) }.to yield_successive_args(
+                Metasploit::Framework::Credential.new(public: "adsf", private: "password"),
+                Metasploit::Framework::Credential.new(public: "jkl", private: "password"),
+                Metasploit::Framework::Credential.new(public: "test_public", private: "password"),
+              )
+            end
+          end
+        end
+      end
+    end
+
+    context "when using password spraying" do
+      let(:password_spray) { true }
+      let(:username) { nil }
+      let(:password) { nil }
+
+      context "when :blank_passwords is true" do
+        let(:blank_passwords) { true }
+
+        context "with password (but no username)" do
+          let(:password) { "pass" }
+
+          # REMOVE BEFORE COMMIT: this yields empty creds (no username, no pass)
+          specify  do
+            expect { |b| collection.each(&b) }.to yield_successive_args()
+          end
+        end
+
+        # REMOVE BEFORE COMMIT: yields nothings, fails because of bug in method
+        context "with username (but no password)" do
+          let(:username) { "user" }
+
+          specify  do
+            expect { |b| collection.each(&b) }.to yield_successive_args(
+              Metasploit::Framework::Credential.new(public: username, private: ''),
+            )
+          end
+        end
+
+        context "when given a user_file" do
+          let(:user_file) do
+            filename = "foo"
+            stub_file = StringIO.new("asdf\njkl\n")
+            allow(File).to receive(:open).with(filename,/^r/).and_return stub_file
+
+            filename
+          end
+
+          # REMOVE BEFORE COMMIT: yields nothing, same for blank passwords option
+          specify  do
+            expect { |b| collection.each(&b) }.to yield_successive_args(
+              Metasploit::Framework::Credential.new(public: "asdf", private: ''),
+              Metasploit::Framework::Credential.new(public: "jkl", private: ''),
+            )
+          end
+        end
+      end
+    end
+
+    context "when every possible option is used" do
```

Comment:
I tried to create some generic non-regression tests to demonstrate how the credentials should be yielded in case of various user/password data source (in fact in case of all them being active).